pgml-apps/discord-bot/bot.py

```diff
@@ -0,0 +1,164 @@
+            model_name="hkunlp/instructor-xl",
+            model_params={"instruction": "Represent the document for retrieval: "},
+        )
+        return embedding_model_id
+
+    # Register a text splitter to the collection
+    async def register_text_splitter(
+        self, collection, chunk_size: int = 1500, chunk_overlap: int = 40
+    ):
+        splitter_id = await collection.register_text_splitter(
+            splitter_name="RecursiveCharacterTextSplitter",
+            splitter_params={"chunk_size": chunk_size, "chunk_overlap": chunk_overlap},
+        )
+        return splitter_id
+
+    # Run an SQL query and return the result
+    async def run_query(self, statement: str, sql_params: tuple = None):
+        conn = await self.pool.getconn()
+        cur = conn.cursor()
+        try:
+            await cur.execute(statement, sql_params)
+            return cur.fetchone()
+        except Exception as e:
+            print(e)
+        finally:
+            cur.close()
+
+    # Query a collection with a string and return vector search results
+    async def query_collection(self, collection_name: str, query: str):
+        collection = await self.pgml.create_or_get_collection(collection_name)
+        return collection.vector_search(
+            query,
+            top_k=3,
+            model_id=2,
+            splitter_id=2,
+            query_parameters={
+                "instruction": "Represent the question for retrieving supporting documents: "
+            },
+        )
+
+    # Start the Discord bot, listen to messages in 'bot-testing' channel and handle the messages
+    async def start(self, collection_name: str, discord_token: str, channel_name: str):
+        self.discord_token = discord_token
+        self.discord_client = self.init_discord_client()
+
+        @self.discord_client.event
+        async def on_ready():
+            print(f"We have logged in as {self.discord_client.user}")
+
+        @self.discord_client.event
+        async def on_message(message):
+            print(f"Message from {message.author}: {message.content}")
+
+            if (
+                message.author != self.discord_client.user
+                and message.channel.name == channel_name
+            ):
+                await self.handle_message(collection_name, message)
+
+        self.discord_client.run(self.discord_token)
+
+    # Handle incoming messages, perform a search on the collection, and respond with a generated answer
+    async def handle_message(self, collection_name, message):
+        print(f"Message from {message.author}: {message.content}")
+        print("Searching the vector database")
+        res = await self.query_collection(collection_name, message.content)
+        print(f"Found {len(res)} results")
+        context = await self.build_context(res)
+        print("Running Completion query")
+        completion = await self.run_transform_sql(context, message.content)
```

Comment:
The SDK should expose a raw connection or give the ability to execute arbitrary SQL over the same connection pool as it's using internally.

Reply:
would be helpful